File: MainApp/src/main/java/com/sleticalboy/dailywork/weight/view/CommonPageIndicator.java
package com.sleticalboy.dailywork.weight.view;

import android.content.Context;
import android.graphics.Canvas;
import android.os.Parcel;
import android.os.Parcelable;
import androidx.annotation.Nullable;
import androidx.viewpager.widget.ViewPager;
import android.util.AttributeSet;
import android.view.View;

/**
 * Created on 18-3-15.
 *
 * @author sleticalboy
 * @description 圆点页面指示器
 */
public class CommonPageIndicator extends View implements PageIndicator {

    private static final int INVALID_POINTER = -1;

    private int mRadius;
    private int mCurrentPage;
    private ViewPager mViewPager;
    private ViewPager.OnPageChangeListener mInternalListener;

    public CommonPageIndicator(Context context) {
        this(context, null);
    }

    public CommonPageIndicator(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public CommonPageIndicator(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init();
    }

    private void init() {
    }

    @Override
    protected void onLayout(boolean changed, int l, int t, int r, int b) {

    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
    }

    @Nullable
    @Override
    protected Parcelable onSaveInstanceState() {
        SavedState savedState = new SavedState(super.onSaveInstanceState());
        savedState.mCurrentPage = mCurrentPage;
        return savedState;
    }

    @Override
    protected void onRestoreInstanceState(Parcelable state) {
        SavedState savedState = (SavedState) state;
        super.onRestoreInstanceState(savedState.getSuperState());
        mCurrentPage = savedState.mCurrentPage;
        requestLayout();
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        setMeasuredDimension(measureWidth(widthMeasureSpec), measureHeight(heightMeasureSpec));
    }

    private int measureWidth(int widthMeasureSpec) {
        int result = 0;
        int specMode = View.MeasureSpec.getMode(widthMeasureSpec);
        int specSize = View.MeasureSpec.getSize(widthMeasureSpec);
        if (specMode == MeasureSpec.EXACTLY) {
            result = specSize;
        } else {
            result = 2 * mRadius + getPaddingTop() + getPaddingBottom() + 1;
            if (specMode == MeasureSpec.AT_MOST) {
                result = Math.min(result, specSize);
            }
        }
        return result;
    }

    private int measureHeight(int heightMeasureSpec) {
        int result = 0;
        int specMode = View.MeasureSpec.getMode(heightMeasureSpec);
        int specSize = View.MeasureSpec.getSize(heightMeasureSpec);
        if (specMode == MeasureSpec.EXACTLY || mViewPager == null || mViewPager.getAdapter() == null) {
            result = specSize;
        } else {
            final int count = mViewPager.getAdapter().getCount();
            result = getPaddingLeft() + getPaddingRight() + 2 * count * mRadius + (count - 1) * mRadius + 1;
            if (specMode == MeasureSpec.AT_MOST) {
                result = Math.min(result, specSize);
            }
        }
        return result;
    }

    static class SavedState extends BaseSavedState {

        public static final Creator<SavedState> CREATOR = new Creator<SavedState>() {
            @Override
            public SavedState createFromParcel(Parcel in) {
                return new SavedState(in);
            }

            @Override
            public SavedState[] newArray(int size) {
                return new SavedState[size];
            }
        };

        int mCurrentPage;

        SavedState(Parcelable superState) {
            super(superState);
        }

        private SavedState(Parcel in) {
            super(in);
            mCurrentPage = in.readInt();
        }

        @Override
        public void writeToParcel(Parcel dest, int flags) {
            super.writeToParcel(dest, flags);
            dest.writeInt(mCurrentPage);
        }
    }

    @Override
    public void setupWithViewPager(ViewPager pagerView, int initialPos) {
        mViewPager = checkValid(mViewPager);
        if (mViewPager.getAdapter() == null) {
            throw new IllegalStateException("没有设置 Adapter");
        }
        mViewPager = pagerView;
        invalidate();
    }

    private <T> T checkValid(T ref) {
        if (ref == null) {
            throw new IllegalArgumentException();
        }
        return ref;
    }

    @Override
    public void setCurrentPage(int pageIndex) {
        mCurrentPage = pageIndex;
        invalidate();
    }

    @Override
    public void notifyDataSetChanged() {
        invalidate();
    }
}

File: MainApp/src/main/java/com/sleticalboy/weight/view/CommonPageIndicator.kt
package com.binlee.weight.view

import android.content.Context
import android.os.Parcel
import android.os.Parcelable
import android.util.AttributeSet
import android.view.View
import androidx.viewpager.widget.ViewPager
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import kotlin.math.min

/**
 * Created on 18-3-15.
 *
 * @author leebin
 * @description 圆点页面指示器
 */
class CommonPageIndicator @JvmOverloads constructor(
  context: Context?, attrs: AttributeSet? = null,
  defStyle: Int = 0
) : View(context, attrs, defStyle), PageIndicator {
  private val mRadius = 0
  private var mCurrentPage = 0
  private var mViewPager: ViewPager? = null
  private val mInternalListener: OnPageChangeListener? = null

  override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {}

  override fun onSaveInstanceState(): Parcelable? {
    val savedState = SavedState(super.onSaveInstanceState())
    savedState.mCurrentPage = mCurrentPage
    return savedState
  }

  override fun onRestoreInstanceState(state: Parcelable) {
    val savedState = state as SavedState
    super.onRestoreInstanceState(savedState.superState)
    mCurrentPage = savedState.mCurrentPage
    requestLayout()
  }

  override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
    setMeasuredDimension(measureWidth(widthMeasureSpec), measureHeight(heightMeasureSpec))
  }

  private fun measureWidth(widthMeasureSpec: Int): Int {
    var result: Int
    val specMode = MeasureSpec.getMode(widthMeasureSpec)
    val specSize = MeasureSpec.getSize(widthMeasureSpec)
    if (specMode == MeasureSpec.EXACTLY) {
      result = specSize
    } else {
      result = 2 * mRadius + paddingTop + paddingBottom + 1
      if (specMode == MeasureSpec.AT_MOST) {
        result = min(result, specSize)
      }
    }
    return result
  }

  private fun measureHeight(heightMeasureSpec: Int): Int {
    var result: Int
    val specMode = MeasureSpec.getMode(heightMeasureSpec)
    val specSize = MeasureSpec.getSize(heightMeasureSpec)
    if (specMode == MeasureSpec.EXACTLY || mViewPager == null || mViewPager!!.adapter == null) {
      result = specSize
    } else {
      val count = mViewPager!!.adapter!!.count
      result = paddingLeft + paddingRight + 2 * count * mRadius + (count - 1) * mRadius + 1
      if (specMode == MeasureSpec.AT_MOST) {
        result = min(result, specSize)
      }
    }
    return result
  }

  internal class SavedState : BaseSavedState {
    var mCurrentPage = 0

    constructor(superState: Parcelable?) : super(superState)
    private constructor(`in`: Parcel) : super(`in`) {
      mCurrentPage = `in`.readInt()
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
      super.writeToParcel(dest, flags)
      dest.writeInt(mCurrentPage)
    }

    companion object {
      @JvmField
      val CREATOR: Parcelable.Creator<SavedState?> = object : Parcelable.Creator<SavedState?> {
        override fun createFromParcel(`in`: Parcel): SavedState? {
          return SavedState(`in`)
        }

        override fun newArray(size: Int): Array<SavedState?> {
          return arrayOfNulls(size)
        }
      }
    }
  }

  override fun setupWithViewPager(viewPager: ViewPager?, initialPos: Int) {
    mViewPager = checkValid(mViewPager)
    checkNotNull(mViewPager!!.adapter) { "没有设置 Adapter" }
    mViewPager = viewPager
    invalidate()
  }

  private fun <T> checkValid(ref: T): T {
    return ref
  }

  override fun setCurrentPage(pageIndex: Int) {
    mCurrentPage = pageIndex
    invalidate()
  }

  override fun notifyDataSetChanged() {
    invalidate()
  }

  init {
  }
}